File: Sources/BraintreeApplePay/BTApplePayClient.m
#import "BTApplePayClient_Internal.h"
#import <PassKit/PassKit.h>

// MARK: - Objective-C File Imports for Package Managers
#if __has_include(<Braintree/BraintreeApplePay.h>) // CocoaPods
#import <Braintree/BTApplePayCardNonce.h>

#elif SWIFT_PACKAGE // SPM
#import <BraintreeApplePay/BTApplePayCardNonce.h>

#else // Carthage
#import <BraintreeApplePay/BTApplePayCardNonce.h>
#endif

// MARK: - Swift File Imports for Package Managers
#if __has_include(<Braintree/Braintree-Swift.h>) // CocoaPods
#import <Braintree/Braintree-Swift.h>

#elif SWIFT_PACKAGE                              // SPM
/* Use @import for SPM support
 * See https://forums.swift.org/t/using-a-swift-package-in-a-mixed-swift-and-objective-c-project/27348
 */
@import BraintreeCore;

#elif __has_include("Braintree-Swift.h")         // CocoaPods for ReactNative
/* Use quoted style when importing Swift headers for ReactNative support
 * See https://github.com/braintree/braintree_ios/issues/671
 */
#import "Braintree-Swift.h"

#else                                            // Carthage
#import <BraintreeCore/BraintreeCore-Swift.h>
#endif

NSString *const BTApplePayErrorDomain = @"com.braintreepayments.BTApplePayErrorDomain";

@implementation BTApplePayClient

#pragma mark - Initialization

- (instancetype)initWithAPIClient:(BTAPIClient *)apiClient {
    if (self = [super init]) {
        _apiClient = apiClient;
    }
    return self;
}

- (instancetype)init {
    return nil;
}

#pragma mark - Public methods

- (void)paymentRequest:(void (^)(PKPaymentRequest * _Nullable, NSError * _Nullable))completion {
    if (!self.apiClient) {
        NSError *error = [NSError errorWithDomain:BTApplePayErrorDomain
                                             code:BTApplePayErrorTypeIntegration
                                         userInfo:@{NSLocalizedDescriptionKey: @"BTAPIClient is nil."}];
        [self invokeBlock:completion onMainThreadWithPaymentRequest:nil error:error];
        return;
    }

    [self.apiClient fetchOrReturnRemoteConfiguration:^(BTConfiguration * _Nullable configuration, NSError * _Nullable error) {
        if (error) {
            [self.apiClient sendAnalyticsEvent:@"ios.apple-pay.error.configuration"];
            [self invokeBlock:completion onMainThreadWithPaymentRequest:nil error:error];
            return;
        }

        if (!configuration.isApplePayEnabled) {
            NSError *error = [NSError errorWithDomain:BTApplePayErrorDomain
                                                 code:BTApplePayErrorTypeUnsupported
                                             userInfo:@{ NSLocalizedDescriptionKey: @"Apple Pay is not enabled for this merchant. Please ensure that Apple Pay is enabled in the control panel and then try saving an Apple Pay payment method again." }];
            [self invokeBlock:completion onMainThreadWithPaymentRequest:nil error:error];
            [self.apiClient sendAnalyticsEvent:@"ios.apple-pay.error.disabled"];
            return;
        }

        PKPaymentRequest *paymentRequest = [[PKPaymentRequest alloc] init];
        paymentRequest.countryCode = configuration.applePayCountryCode;
        paymentRequest.currencyCode = configuration.applePayCurrencyCode;
        paymentRequest.merchantIdentifier = configuration.applePayMerchantIdentifier;
        paymentRequest.supportedNetworks = configuration.applePaySupportedNetworks;
        
        [self invokeBlock:completion onMainThreadWithPaymentRequest:paymentRequest error:nil];
    }];
}

- (void)tokenizeApplePayPayment:(PKPayment *)payment completion:(void (^)(BTApplePayCardNonce *, NSError *))completionBlock {
    if (!self.apiClient) {
        NSError *error = [NSError errorWithDomain:BTApplePayErrorDomain
                                             code:BTApplePayErrorTypeIntegration
                                         userInfo:@{NSLocalizedDescriptionKey: @"BTApplePayClient tokenization failed because BTAPIClient is nil."}];
        completionBlock(nil, error);
        return;
    }
    
    [self.apiClient sendAnalyticsEvent:@"ios.apple-pay.start"];
   
    if (!payment) {
        NSError *error = [NSError errorWithDomain:BTApplePayErrorDomain
                                             code:BTApplePayErrorTypeUnsupported
                                         userInfo:@{NSLocalizedDescriptionKey: @"A valid PKPayment is required."}];
        completionBlock(nil, error);
        [self.apiClient sendAnalyticsEvent:@"ios.apple-pay.error.invalid-payment"];
        return;
    }
    
    [self.apiClient fetchOrReturnRemoteConfiguration:^(BTConfiguration *configuration, NSError *error) {
        if (error) {
            [self.apiClient sendAnalyticsEvent:@"ios.apple-pay.error.configuration"];
            completionBlock(nil, error);
            return;
        }

        if (![configuration.json[@"applePay"][@"status"] isString] ||
            [[configuration.json[@"applePay"][@"status"] asString] isEqualToString:@"off"]) {
            NSError *error = [NSError errorWithDomain:BTApplePayErrorDomain
                                                 code:BTApplePayErrorTypeUnsupported
                                             userInfo:@{ NSLocalizedDescriptionKey: @"Apple Pay is not enabled for this merchant. Please ensure that Apple Pay is enabled in the control panel and then try saving an Apple Pay payment method again." }];
            completionBlock(nil, error);
            [self.apiClient sendAnalyticsEvent:@"ios.apple-pay.error.disabled"];
            return;
        }
        
        NSMutableDictionary *parameters = [NSMutableDictionary new];
        parameters[@"applePaymentToken"] = [self parametersForPaymentToken:payment.token];
        parameters[@"_meta"] = @{
            @"source" : self.apiClient.metadata.sourceString,
            @"integration" : self.apiClient.metadata.integrationString,
            @"sessionId" : self.apiClient.metadata.sessionID,
        };
        
        [self.apiClient POST:@"v1/payment_methods/apple_payment_tokens"
                  parameters:parameters
                  completion:^(BTJSON *body, __unused NSHTTPURLResponse *response, NSError *error) {
            if (error) {
                if (error.code == BTCoreConstants.networkConnectionLostCode) {
                    [self.apiClient sendAnalyticsEvent:@"ios.apple-pay.network-connection.failure"];
                }
                completionBlock(nil, error);
                [self.apiClient sendAnalyticsEvent:@"ios.apple-pay.error.tokenization"];
                return;
            }

            BTApplePayCardNonce *tokenized = [[BTApplePayCardNonce alloc] initWithJSON:body[@"applePayCards"][0]];

            completionBlock(tokenized, nil);
            [self.apiClient sendAnalyticsEvent:@"ios.apple-pay.success"];
        }];
    }];
}

#pragma mark - Helpers

- (NSDictionary *)parametersForPaymentToken:(PKPaymentToken *)token {
    NSMutableDictionary *mutableParameters = [NSMutableDictionary dictionary];

    mutableParameters[@"paymentData"] = [token.paymentData base64EncodedStringWithOptions:0];
    mutableParameters[@"transactionIdentifier"] = token.transactionIdentifier;
    mutableParameters[@"paymentInstrumentName"] = token.paymentMethod.displayName;
    mutableParameters[@"paymentNetwork"] = token.paymentMethod.network;

    return [mutableParameters copy];
}

- (void)invokeBlock:(nonnull void (^)(PKPaymentRequest * _Nullable, NSError * _Nullable))completion onMainThreadWithPaymentRequest:(nullable PKPaymentRequest *)paymentRequest error:(nullable NSError *)error {
    dispatch_async(dispatch_get_main_queue(), ^{
        completion(paymentRequest, error);
    });
}

@end

File: Sources/BraintreeApplePay/BTApplePayClient.swift
import Foundation
import PassKit

#if canImport(BraintreeCore)
import BraintreeCore
#endif

/// Used to process Apple Pay payments
@objc public class BTApplePayClient: NSObject {

    // MARK: - Internal Properties

    /// Exposed for testing to get the instance of BTAPIClient
    var apiClient: BTAPIClient

    // MARK: - Initializer

    /// Creates an Apple Pay client
    /// - Parameter authorization: A client token or tokenization key
    @objc(initWithAuthorization:)
    public init(authorization: String) {
        self.apiClient = BTAPIClient(authorization: authorization)
    }

    // MARK: - Public Methods

    /// Creates a `PKPaymentRequest` with values from your Braintree Apple Pay configuration.
    /// It populates the following values of `PKPaymentRequest`: `countryCode`, `currencyCode`, `merchantIdentifier`, `supportedNetworks`.
    /// - Parameter completion: A completion block that returns the payment request or an error.
    @objc(makePaymentRequest:)
    public func makePaymentRequest(completion: @escaping (PKPaymentRequest?, Error?) -> Void) {
        Task { @MainActor in
            do {
                let paymentRequest = try await makePaymentRequest()
                completion(paymentRequest, nil)
            } catch {
                completion(nil, error)
            }
        }
    }

    /// Creates a `PKPaymentRequest` with values from your Braintree Apple Pay configuration.
    /// It populates the following values of `PKPaymentRequest`: `countryCode`, `currencyCode`, `merchantIdentifier`, `supportedNetworks`.
    /// - Returns: A `PKPaymentRequest`
    /// - Throws: An `Error` describing the failure
    public func makePaymentRequest() async throws -> PKPaymentRequest {
        apiClient.sendAnalyticsEvent(BTApplePayAnalytics.paymentRequestStarted)

        do {
            let configuration = try await apiClient.fetchOrReturnRemoteConfiguration()

            guard configuration.isApplePayEnabled else {
                sendPaymentRequestFailureAnalytics(with: BTApplePayError.unsupported)
                throw BTApplePayError.unsupported
            }

            let paymentRequest = PKPaymentRequest()
            paymentRequest.countryCode = configuration.applePayCountryCode ?? ""
            paymentRequest.currencyCode = configuration.applePayCurrencyCode ?? ""
            paymentRequest.merchantIdentifier = configuration.applePayMerchantIdentifier ?? ""
            paymentRequest.supportedNetworks = configuration.applePaySupportedNetworks ?? []

            return notifyPaymentRequestSuccess(with: paymentRequest)
        } catch {
            sendPaymentRequestFailureAnalytics(with: error)
            throw error
        }
    }

    /// Checks if Apple Pay is configured and available for the current merchant account and device.
    /// - Parameter completion: A completion block that returns `true` if Apple Pay is supported for the customer.
    @objc(isApplePaySupported:)
    public func isApplePaySupported(completion: @escaping (Bool) -> Void) {
        Task { @MainActor in
            let isSupported = await isApplePaySupported()
            completion(isSupported)
        }
    }

    /// Checks if Apple Pay is configured and available for the current merchant account and device.
    /// - Returns: A `Bool` that returns true if Apple Pay is supported for the customer.
    public func isApplePaySupported() async -> Bool {
        do {
            let configuration = try await apiClient.fetchOrReturnRemoteConfiguration()
            return configuration.canMakeApplePayPayments
        } catch {
            return false
        }
    }

    /// Tokenizes an Apple Pay payment.
    /// - Parameters:
    ///   - payment: A `PKPayment` instance, typically obtained by presenting a `PKPaymentAuthorizationViewController`
    ///   - completion: A completion block that is invoked when tokenization has completed. If tokenization succeeds, we will return a `BTApplePayCardNonce`
    ///   and `error` will be `nil`; if it fails, `BTApplePayCardNonce` will be `nil` and `error` will describe the failure.
    @objc(tokenizeApplePayPayment:completion:)
    public func tokenize(_ payment: PKPayment, completion: @escaping (BTApplePayCardNonce?, Error?) -> Void) {
        Task { @MainActor in
            do {
                let applePayNonce = try await tokenize(payment)
                completion(applePayNonce, nil)
            } catch {
                completion(nil, error)
            }
        }
    }

    /// Tokenizes an Apple Pay payment.
    /// - Parameter payment: A `PKPayment` instance, typically obtained by presenting a `PKPaymentAuthorizationViewController`
    /// - Returns: A `BTApplePayCardNonce`
    /// - Throws: An `Error` describing the failure
    public func tokenize(_ payment: PKPayment) async throws -> BTApplePayCardNonce {
        apiClient.sendAnalyticsEvent(BTApplePayAnalytics.tokenizeStarted)

        do {
            let configuration = try await apiClient.fetchOrReturnRemoteConfiguration()

            guard configuration.isApplePayEnabled else {
                sendTokenizeFailureAnalytics(with: BTApplePayError.unsupported)
                throw BTApplePayError.unsupported
            }

            let parameters = BTApplePaymentTokensRequest(token: payment.token)
            let (body, _) = try await apiClient.post("v1/payment_methods/apple_payment_tokens", parameters: parameters)

            guard let body else {
                sendTokenizeFailureAnalytics(with: BTApplePayError.noApplePayCardsReturned)
                throw BTApplePayError.noApplePayCardsReturned
            }

            guard let applePayNonce = BTApplePayCardNonce(json: body["applePayCards"][0]) else {
                sendTokenizeFailureAnalytics(with: BTApplePayError.failedToCreateNonce)
                throw BTApplePayError.failedToCreateNonce
            }

            return notifyTokenizeSuccess(with: applePayNonce)
        } catch {
            sendTokenizeFailureAnalytics(with: error)
            throw error
        }
    }
    
    // MARK: - Analytics Helper Methods

    private func notifyTokenizeSuccess(with result: BTApplePayCardNonce) -> BTApplePayCardNonce {
        apiClient.sendAnalyticsEvent(BTApplePayAnalytics.tokenizeSucceeded)
        return result
    }

    private func sendTokenizeFailureAnalytics(with error: Error) {
        apiClient.sendAnalyticsEvent(BTApplePayAnalytics.tokenizeFailed, errorDescription: error.localizedDescription)
    }

    private func notifyPaymentRequestSuccess(with result: PKPaymentRequest) -> PKPaymentRequest {
        apiClient.sendAnalyticsEvent(BTApplePayAnalytics.paymentRequestSucceeded)
        return result
    }

    private func sendPaymentRequestFailureAnalytics(with error: Error) {
        apiClient.sendAnalyticsEvent(BTApplePayAnalytics.paymentRequestFailed, errorDescription: error.localizedDescription)
    }
}
